File: python/array_and_linkedlist/my_list.py
class MyList:
    def __init__(self):
        self._capacity: int = 10  # 容量
        self._arr: list[int] = [0] * self._capacity  # 数组
        self._size: int = 0
        self._extend_ratio = 2  # 每次列表扩容的倍数

    def size(self) -> int:
        return self._size

    def capacity(self) -> int:
        return self._capacity

    def get(self, index: int) -> int:
        self.check_index(index)
        return self._arr[index]

    def set(self, index: int, value: int) -> None:
        self.check_index(index)
        self._arr[index] = value

    def add(self, value: int) -> None:
        if self.size() == self.capacity():
            self.extend_ratio()
        self._arr[self._size] = value
        self._size += 1

    def insert(self, index: int, value: int) -> None:
        # 容量满了需要先扩容
        self.check_index(index)
        if self.size() == self.capacity():
            self.extend_ratio()
        for i in range(self.size() - 1, index - 1, -1):
            self._arr[i+1] = self._arr[i]
        self._arr[index] = value
        self._size += 1

    def remove(self, index: int) -> None:
        self.check_index(index)
        for i in range(index, self.size() - 1):
            self._arr[i] = self._arr[i+1]
        self._size -= 1

    def extend_ratio(self):
        self._arr = self._arr + [0] * self.capacity() * (self._extend_ratio - 1)
        self._capacity = len(self._arr)

    def to_array(self) -> list:
        return self._arr[:self._size]

    def check_index(self, index: int) -> None:
        if index < 0 or index >= self.size():
            raise IndexError("索引越界")


def main():
    # 初始化列表
    nums = MyList()
    # 在尾部添加元素
    nums.add(1)
    nums.add(3)
    nums.add(2)
    nums.add(5)
    nums.add(4)
    print(f"列表 nums = {nums.to_array()} ，容量 = {nums.capacity()} ，长度 = {nums.size()}")

    # 在中间插入元素
    nums.insert(3, 6)
    print("在索引 3 处插入数字 6 ，得到 nums =", nums.to_array())

    # 删除元素
    nums.remove(3)
    print("删除索引 3 处的元素，得到 nums =", nums.to_array())

    # 访问元素
    num = nums.get(1)
    print("访问索引 1 处的元素，得到 num =", num)

    # 更新元素
    nums.set(1, 0)
    print("将索引 1 处的元素更新为 0 ，得到 nums =", nums.to_array())

    # 测试扩容机制
    for i in range(10):
        # 在 i = 5 时，列表长度将超出列表容量，此时触发扩容机制
        nums.add(i)
    print(f"扩容后的列表 {nums.to_array()} ，容量 = {nums.capacity()} ，长度 = {nums.size()}")


if __name__ == "__main__":
    main()

File: go/array_and_linkedlist/my_list.go
/*
基于 slice 实现一个列表
*/
package main

import "fmt"

/* 列表结构体 */
type myList struct {
    arrCapacity int
    arr         []int
    arrSize     int
    extendRatio int
}

/* 构造函数 */
func newMyList() *myList {
    return &myList{
        arrCapacity: 10,              // 列表容量
        arr:         make([]int, 10), // 数组
        arrSize:     0,               // 列表长度
        extendRatio: 2,               // 扩容倍数
    }
}

/* 获取长度 */
func (l *myList) size() int {
    return l.arrSize
}

/* 获取容量 */
func (l *myList) capacity() int {
    return l.arrCapacity
}

/* 访问元素 */
func (l *myList) get(index int) int {
    l.checkIndex(index)
    return l.arr[index]

}

/* 更新元素 */
func (l *myList) set(index int, num int) {
    l.checkIndex(index)
    l.arr[index] = num
}

/* 在尾部添加元素 */
func (l *myList) add(num int) {
    l.checkIsNeedExtendCapacity()
    l.arr[l.arrSize] = num
    l.arrSize++
}

/* 在中间插入元素 */
func (l *myList) insert(index int, num int) {
    l.checkIndex(index)
    l.checkIsNeedExtendCapacity()
    for i := l.size() - 1; i >= index; i-- {
        l.arr[i+1] = l.arr[i]
    }
    l.arr[index] = num
    l.arrSize++

}

/* 删除元素 */
func (l *myList) remove(index int) int {
    l.checkIndex(index)
    num := l.arr[index]
    for i := index; i < l.size()-1; i++ {
        l.arr[i] = l.arr[i+1]
    }
    l.arrSize--
    return num
}

/* 返回有效长度的列表 */
func (l *myList) toArray() []int {
    return l.arr[:l.size()]
}

/* 检查是否需要扩容 */
func (l *myList) checkIsNeedExtendCapacity() {
    if l.size() == l.capacity() {
        l.extendCapacity()
    }
}

/* 扩容 */
func (l *myList) extendCapacity() {
    l.arr = append(l.arr, make([]int, l.arrCapacity*(l.extendRatio-1))...)
    l.arrCapacity = len(l.arr)
}

/* 检查索引 */
func (l *myList) checkIndex(index int) {
    if index < 0 || index >= l.size() {
        panic("索引越界")
    }
}

func main() {
    nums := newMyList()
    /* 在尾部添加元素 */
    nums.add(1)
    nums.add(3)
    nums.add(2)
    nums.add(5)
    nums.add(4)
    fmt.Printf("列表 nums = %v ，容量 = %v ，长度 = %v\n", nums.toArray(), nums.capacity(), nums.size())

    /* 在中间插入元素 */
    nums.insert(3, 6)
    fmt.Printf("在索引 3 处插入数字 6 ，得到 nums = %v\n", nums.toArray())

    /* 删除元素 */
    nums.remove(3)
    fmt.Printf("删除索引 3 处的元素，得到 nums = %v\n", nums.toArray())

    /* 访问元素 */
    num := nums.get(1)
    fmt.Printf("访问索引 1 处的元素，得到 num = %v\n", num)

    /* 更新元素 */
    nums.set(1, 0)
    fmt.Printf("将索引 1 处的元素更新为 0 ，得到 nums = %v\n", nums.toArray())

    /* 测试扩容机制 */
    for i := 0; i < 10; i++ {
        // 在 i = 5 时，列表长度将超出列表容量，此时触发扩容机制
        nums.add(i)
    }
    fmt.Printf("扩容后的列表 nums = %v ，容量 = %v ，长度 = %v\n", nums.toArray(), nums.capacity(), nums.size())
}
